### PR TITLE
feat: building internal state management tool

### LINKED ISSUE #31: Test building internal state management tool
Build a state management system for internal usage. With `proxies`, this could be integrated with socket types

### PR DESCRIPTION
**Pull request by o0victor0o**
- [x] Look over your code one last time
- [x] Manually test locally all functionality and user interface changes in the browser you expect to be QA'd
- [x] Describe what changes this PR includes
- [x] Post screenshots/gifs for design review, applicable if there are user-facing UI/UX changes
- [x] Fill in QA Notes listing clear, ordered steps for peer, pre-merge QA on the PR

## Description

Proxy utility class.
This class can proxy a given object - this means that properties of the object that are read or set can be intercepted with an event dispatch.

When a property is set, if a set event was registered on the proxy, it will be triggered with the old and new values.
When a property is read, if a read event was registered, it will be called to get the return value (this is useful for nested proxies - see the second test case)

Closes #31.

## Screenshots/Gifs

[add screenshots/gifs here]

## QA Notes

Tested the case where changing a shallow object's field is proxied.
Tested the case where changing a nested object's field is proxied

PR Author Browser Info:

```
**QA Tester Browser Info:** 

## Testing Steps

- [ ] [Add QA steps here]

## Notes

- [ ] [Tester should add any comments, questions, or concerns that pop up during QA]
```

### CODE REVIEW

File: packages/shared/src/util/proxyable.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import _ from 'lodash';
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts:1]: We should split up our lodash imports to decrease our bundle size
```
import isNil from 'lodash/isNil';
...
```

File: packages/shared/src/util/proxyable.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+type PropertyChangeEvent<T> = (oldValue: T, newValue: T) => void;
+type PropertyReadEvent<T> = (value: T) => T;
+
+/**
+ * @class
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts:7]: Love the JS doc, great comments!

File: packages/shared/src/util/proxyable.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+   * @param key The key to set the event for.
+   * @param event The event to trigger when the property is changed.
+   */
+  public onPropertyChange<TKey extends keyof T>(
+    key: TKey,
+    event: PropertyChangeEvent<T[TKey]>,
+  ): Proxyable<T> {
+    this.ChangeEvents.set(
+      key.toString(),
+      event as PropertyChangeEvent<T[keyof T]>,
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts:32]: not sure why we would need a cast here, looks like `event: PropertyChangeEvent<T[TKey]>` is the same type as listed

File: packages/shared/src/util/proxyable.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  constructor(obj: T) {
+    this.ProxyObj = obj;
+  }
+
+  /**
+   * Sets a property change event for a given key on the object.
+   * @param key The key to set the event for.
+   * @param event The event to trigger when the property is changed.
+   */
+  public onPropertyChange<TKey extends keyof T>(
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts:26]: *nit we should opt for arrow functions even within classes. Maybe a topic of discussion (or a best practises doc), but ideally we keep it consistent across the board. Ill enable a eslint rule once we decide

File: packages/shared/src/util/proxyable.ts (hunk 3 above)
Comment (o0victor0o) [reply packages/shared/src/util/proxyable.ts:32]: That's what I thought but it gives a type error I couldn't work around

File: packages/shared/src/util/proxyable.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   * @param key The key to set the event for.
+   * @param event The event to trigger when the property is read.
+   */
+  public onPropertyRead<TKey extends keyof T>(
+    key: TKey,
+    event: PropertyReadEvent<T[TKey]>,
+  ): Proxyable<T> {
+    this.ReadEvents.set(
+      key.toString(),
+      event as unknown as PropertyReadEvent<T[keyof T]>,
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts:48]: Same here, not sure why we would need to cast to unknown here

File: packages/shared/src/util/proxyable.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    return true;
+  }
+
+  /**
+   * The proxy get handler.
+   * If a read event exists for the property, it will be called and the result returned.
+   * Otherwise, the value of the property will be returned.
+   */
+  private proxyGet(obj: T, key: keyof T): unknown {
```
Comment (AydanGaite) [inline packages/shared/src/util/proxyable.ts]: Ideally we have a better return type than unknown. We can at least determine that `undefined` is a possibility

File: packages/shared/src/util/__tests__/proxyable.test.ts
Comment (AydanGaite) [inline packages/shared/src/util/__tests__/proxyable.test.ts:1]: Tests look good!

File: packages/shared/src/util/proxyable.ts (hunk 3 above)
Comment (AydanGaite) [reply packages/shared/src/util/proxyable.ts:32]: There should be a way, id be happy to pair with you to try to come up with a solution/workaround. Ideally we cast as little as possible

**Review by AydanGaite — APPROVED** [review]

**Pull request merged**